diff -x '*~' -x .git -urw --new-file finpar/LocVolCalib/AllParOpenCLMP/Vect_GPU.h finpar-2/LocVolCalib/AllParOpenCLMP/Vect_GPU.h
--- finpar/LocVolCalib/AllParOpenCLMP/Vect_GPU.h	2022-03-25 17:48:03.783167000 +0100
+++ finpar-2/LocVolCalib/AllParOpenCLMP/Vect_GPU.h	2022-03-25 18:19:49.480007000 +0100
@@ -155,6 +155,9 @@
 
     unsigned long int elapsed;
     struct timeval t_start, t_end, t_diff;
+
+    clFinish(cqCommandQueue[dev_id]);
+
     gettimeofday(&t_start, NULL);
     
     // now execute kernels and record the time!
@@ -162,6 +165,8 @@
             run_GPUkernels_one_time_iteration ( cqCommandQueue[dev_id], kernels );
     } // END TIME LOOP!
 
+    clFinish(cqCommandQueue[dev_id]);
+    gettimeofday(&t_end, NULL);
 
     { // WRITE BACK THE RESULT ARRAY TO CPU !!! //
         cl_int  ciErr;
@@ -178,7 +183,6 @@
             );
     }
 
-    gettimeofday(&t_end, NULL);
     timeval_subtract(&t_diff, &t_end, &t_start);
     elapsed = t_diff.tv_sec*1e6+t_diff.tv_usec;
 
diff -x '*~' -x .git -urw --new-file finpar/LocVolCalib/includeC/Constants.h finpar-2/LocVolCalib/includeC/Constants.h
--- finpar/LocVolCalib/includeC/Constants.h	2022-03-25 18:17:44.503327000 +0100
+++ finpar-2/LocVolCalib/includeC/Constants.h	2022-03-25 18:17:38.972936000 +0100
@@ -18,7 +18,7 @@
 #define WARP            (1<<lgWARP) 
 
 #define BLOCK_DIM           16
-#define logWORKGROUP_SIZE   8
+#define logWORKGROUP_SIZE   9
 #define    WORKGROUP_SIZE   (1<<logWORKGROUP_SIZE) 
     
 typedef struct {
diff -x '*~' -x .git -urw --new-file finpar/LocVolCalib/includeC/ParseInput.h finpar-2/LocVolCalib/includeC/ParseInput.h
--- finpar/LocVolCalib/includeC/ParseInput.h	2022-03-25 18:17:44.510841000 +0100
+++ finpar-2/LocVolCalib/includeC/ParseInput.h	2022-03-25 18:17:38.975701000 +0100
@@ -59,7 +59,7 @@
         assert(atr_ok && "Illegal NUM_X value!");
 
         atr_ok  = (num_Y > 0) && (num_Y <= WORKGROUP_SIZE) && is_pow2(num_Y); 
-        assert(atr_ok && "Illegal NUM_X value!");
+        assert(atr_ok && "Illegal NUM_Y value!");
 
         atr_ok  = num_T > 0;
         assert(atr_ok && "NUM_T value less or equal to zero!!");
@@ -135,4 +135,3 @@
 }
 #endif
 #endif // PARSE_INPUT
-
diff -x '*~' -x .git -urw --new-file finpar/LocVolCalib/OutParOpenCLMP/Vect_GPU.h finpar-2/LocVolCalib/OutParOpenCLMP/Vect_GPU.h
--- finpar/LocVolCalib/OutParOpenCLMP/Vect_GPU.h	2022-03-25 17:48:03.889980000 +0100
+++ finpar-2/LocVolCalib/OutParOpenCLMP/Vect_GPU.h	2022-03-25 18:20:03.154003000 +0100
@@ -145,12 +145,18 @@
     unsigned long int elapsed;
     { // now execute kernels!
         struct timeval t_start, t_end, t_diff;
+
+        clFinish(cqCommandQueue[dev_id]);
+
         gettimeofday(&t_start, NULL);
 
         for(int t_ind = NUM_T-2; t_ind>=0; --t_ind) {
             run_GPUkernels_one_time_iteration ( cqCommandQueue[dev_id], kernels );
         } // END TIME LOOP!
 
+        clFinish(cqCommandQueue[dev_id]);
+
+        gettimeofday(&t_end, NULL);
 
         { // WRITE BACK THE RESULT ARRAY TO CPU !!! //
             cl_int  ciErr;
@@ -167,7 +173,6 @@
                 );
         }
 
-        gettimeofday(&t_end, NULL);
         timeval_subtract(&t_diff, &t_end, &t_start);
         elapsed = t_diff.tv_sec*1e6+t_diff.tv_usec;
     }
diff -x '*~' -x .git -urw --new-file finpar/OptionPricing/CppOpenCL/GenPricing.cpp finpar-2/OptionPricing/CppOpenCL/GenPricing.cpp
--- finpar/OptionPricing/CppOpenCL/GenPricing.cpp	2022-03-25 17:48:03.995425000 +0100
+++ finpar-2/OptionPricing/CppOpenCL/GenPricing.cpp	2022-03-25 18:20:58.452797000 +0100
@@ -74,6 +74,9 @@
 
     // we do not measure the just-in-time compilation time!
     struct timeval t_start, t_end, t_diff;
+
+    clFinish(cqCommandQueue[GPU_DEV_ID]);
+
     gettimeofday(&t_start, NULL);
 
     if(kernel_type == PRIV) { // CALL THE PRIVATE KERNEL
@@ -128,6 +131,8 @@
         ro_scal.sobol_count_ini = sob_ini_count;
     }
 
+    clFinish(cqCommandQueue[GPU_DEV_ID]);
+
     gettimeofday(&t_end, NULL);
     timeval_subtract(&t_diff, &t_end, &t_start);
     elapsed = t_diff.tv_sec*1e6+t_diff.tv_usec;
diff -x '*~' -x .git -urw --new-file finpar/platform.mk finpar-2/platform.mk
--- finpar/platform.mk	2022-03-25 18:17:44.571252000 +0100
+++ finpar-2/platform.mk	2022-03-25 18:17:38.458105000 +0100
@@ -14,11 +14,11 @@
 GPU_CONST_MEM = 64
 GPU_REG_MEM   = 64
 # device memory in gigabyte
-GPU_DEVICE_MEM= 2
+GPU_DEVICE_MEM= 40
 # ``Optimal'' Amount of Local/Fast Memory Per Thread 
 GPU_LOCAL_MEM_PER_TH=8
 # Number of GPU cores
-GPU_NUM_CORES = 1536
+GPU_NUM_CORES = 6912
 
 # CPU and memory spec.
 NCORES = 32
